consensus/merge/src/main/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinator.java

```diff
@@ -208,75 +216,119 @@ public PayloadIdentifier preparePayload(
+
+    while (!isBlockCreationCancelled(payloadIdentifier)) {
+      try {
+        recoverableBlockCreation(payloadIdentifier, blockCreator, System.currentTimeMillis());
+      } catch (final CancellationException ce) {
+        debugLambda(
+            LOG,
+            "Block creation for payload id {} has been cancelled, reason {}",
+            payloadIdentifier::toString,
+            () -> logException(ce));
+        return null;
+      } catch (final Throwable e) {
+        LOG.warn(
+            "Something went wrong creating block for payload id {}, error {}",
+            payloadIdentifier.toHexString(),
+            logException(e));
+        return null;
+      }
+    }
+    return null;
   }
 
-  private CompletableFuture<Block> retryBlockCreation(
+  private void recoverableBlockCreation(
       final PayloadIdentifier payloadIdentifier,
       final Supplier<Block> blockCreator,
-      final long remainingTime) {
-    final long startedAt = System.currentTimeMillis();
+      final long startedAt) {
+
+    try {
+      newBlockFound(blockCreator.get(), payloadIdentifier, startedAt);
+    } catch (final Throwable throwable) {
+      if (canRetryBlockCreation(throwable) && !isBlockCreationCancelled(payloadIdentifier)) {
+        debugLambda(
+            LOG,
+            "Retrying block creation for payload id {} after recoverable error {}",
+            payloadIdentifier::toHexString,
+            () -> logException(throwable));
+        recoverableBlockCreation(payloadIdentifier, blockCreator, startedAt);
+      } else {
+        throw throwable;
+      }
+    }
+  }
 
-    debugLambda(
-        LOG,
-        "Block creation started for payload id {}, remaining time is {}ms",
-        payloadIdentifier::toShortHexString,
-        () -> remainingTime);
-
-    return exceptionallyCompose(
-        ethContext
-            .getScheduler()
-            .scheduleComputationTask(blockCreator)
-            .orTimeout(remainingTime, TimeUnit.MILLISECONDS),
-        throwable -> {
-          if (canRetryBlockCreation(throwable)) {
-            final long newRemainingTime = remainingTime - (System.currentTimeMillis() - startedAt);
-            debugLambda(
-                LOG,
-                "Retrying block creation for payload id {}, remaining time is {}ms, last error {}",
-                payloadIdentifier::toShortHexString,
-                () -> newRemainingTime,
-                () -> logException(throwable));
-            return retryBlockCreation(payloadIdentifier, blockCreator, newRemainingTime);
-          } else {
-            debugLambda(
-                LOG,
-                "Something went wrong creating block for payload id {}, error {}",
-                payloadIdentifier::toShortHexString,
-                () -> logException(throwable));
-          }
-          return CompletableFuture.failedFuture(throwable);
-        });
+  private void newBlockFound(
```

Comment:
This method is named after the situation being reacted to. Can we rename it so it describes what it is going to do instead? Perhaps upgradeIfTimeLeft?

Reply:
Indeed its name can be improved, and I am thinking about something like `evaluateNewBlock` since it is not yes sure the block is an upgrade, since it still to be validated and the compared with already found blocks, what do you think?